File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

// 곡괭이로 광물을 캘 때 필요한 에너지

int energyNeeded[3][3] = {
    {1, 1, 1},
    {5, 1, 1},
    {25, 5, 1}
};

// mineral string 파싱용 map

unordered_map<string, int> mineralToInt = {
    {"diamond", 0},
    {"iron", 1},
    {"stone", 2}
};

int dfs(int idx, int energy, vector<int> picks, const vector<string>& minerals) {
    // 모든 광물을 캤을 경우
    
    if (idx >= minerals.size()) return energy;
    
    int minEnergy = 987654321;
    
    bool noPicks = true;
    
    for (int i = 0; i < 3; i++) {
        if (!picks[i]) continue;
        
        noPicks = false;
        
        // 광물 5개를 캔다.
        
        picks[i]--;
        int nextEnergy = energy;
        for (int j = idx; j < minerals.size() && j < idx + 5; j++) {
            nextEnergy += energyNeeded[i][mineralToInt[minerals[j]]];
        }
        minEnergy = min(minEnergy,
                        dfs(idx + 5, nextEnergy, picks, minerals));
        picks[i]++;
    }
    
    if (noPicks) return energy;
    
    return minEnergy;
}

int solution(vector<int> picks, vector<string> minerals) {
    int answer = dfs(0, 0, picks, minerals);
    return answer;
}

File: 프로그래머스/2/172927. 광물 캐기/광물 캐기.py
def solution(picks, minerals):
    fatigueTable = [
        [1, 1, 1],
        [5, 1, 1],
        [25, 5, 1]
    ]
    
    mineralDict = {
        'diamond' : 0,
        'iron' : 1,
        'stone' : 2
    }
    
    def parseMineral(mineral):
        return mineralDict[mineral]
    
    def dfs(idx, fatigue, picks, minerals):
        # 모든 광물을 캤을 경우
        
        if idx == len(minerals):
            return fatigue
        
        # 모든 곡괭이를 소모했을 경우
        
        if not sum(picks):
            return fatigue
        
        # 곡괭이 종류별로 DFS
        
        minFatigue = 987654321
        
        for i in range(3):
            if not picks[i]:
                continue
            
            picks[i] -= 1
            
            nextFatigue = fatigue
            for mineral in map(parseMineral, minerals[idx : idx + 5]):
                nextFatigue += fatigueTable[i][mineral]
            
            minFatigue = min(minFatigue, dfs(idx + 5, nextFatigue, picks, minerals))
            
            picks[i] += 1
        
        return minFatigue
            
            
    return dfs(0, 0, picks, minerals)